### PR TITLE
Expt/Hard Abuse Detection

### PR DESCRIPTION
**Pull request by Luc-Mcgrady**
The idea goes as follows (correct me if I'm wrong):

1. w[15] is a measure of the degree to which hard differs from good.
3. If the user is abusing hard, handling hard reviews with `stability_after_failure` should decrease the log loss.
4. Allow w[15] to go negative and use the negative w[15] as a "hard bonus" for `stability_after_failure`.

```py
success = X[:, 1] > (2 - self.w[15])
```
```py
        hard_bonus = torch.where(rating == 2, 3.5 + (self.w[15] * 2.5), 1)
```
(I eyeballed the 3.5, w[15] will be negative so its on a scale of 3.5 - 1 )

Using this script in the srs benchmark: https://gist.github.com/Luc-Mcgrady/89648eb2cdd7233b1267553284c7a22d
I get this result: https://pastebin.com/QnGaWpWv
(noticeably a small log loss improvement overall???!)
(My computers results vary from the regular results for some reason. I ran it both normally and with this change to compare but I still can't trust the results.)
```
---- Abusers: len(values)=63 mean(values)=0.0036330158730158726 sum(values)=0.22887999999999997 ----
```
With users like
```
user=98 w15:-0.9895, before=0.252682, after=0.230099, before-after=0.02258300000000002
user=686 w15:-0.9999, before=0.568046, after=0.536963, before-after=0.03108300000000008
```
Who seem to absolutely be hard abusers provided this method isn't flawed.

Edit: I made a mistake where hards failure stability could be above goods stability, should be fixed now but the values are out of date.

### CODE REVIEW

**Comment by L-M-Sherlock** [comment]
> 2\. If the user is abusing hard, handling hard reviews with `stability_after_failure` should decrease the log loss.

You forget to update the label-mapping. In current implementation, `again` is mapped to zero, and `hard`, `good`, `easy` are mapped to one. If the user abused `hard`, `hard` should be mapped to zero, right?

**Comment by Expertium** [comment]
So how would dynamic labels work? At epoch n we're doing optimization with some labels, then at epoch n+1 we're doing optimization with different labels? That seems like it would cause problems, if even converge at all.

And even if optimization converges and doesn't cause any weird problems, this means that in practice whether Hard counts as "fail" or "pass" could change at any time. The user does a few more reviews and BAM - suddenly, Hard is considered a "fail" now. I really doubt it would lead to good user experience.

**Comment by Luc-Mcgrady** [comment]
> You've forgotten to update the label-mappings. In current implementation, `again` is mapped to zero, while `hard`, `good` and `easy` are mapped to one. If the user abused `hard`, `hard` should be mapped to zero, right?

You're absolutely right but making this change has seemingly drastically negatively affected the log loss. I suppose the current implementation of hard_bonus isn't well implemented.

**Comment by Expertium** [comment]
I tested an idea: making the hard_penalty act as a **multiplier** when it's >0 and as a **weight for a weighted average** when it's <0.

The naive way to let it go below 0 it is like this:
`            w[15] = w[15].clamp(-1, 1)  # Hard penalty`
That's it. No other changes.
However, this creates 2 problems:
1) New S when the user presses Hard could end up being **lower** than post-lapse S aka PLS. Not only that, but it could end up being **negative**. Neither makes sense.
2) Cards with lower difficulty would have **lower** new S and cards with higher difficulty would have **higher** new S, which makes no sense.

So instead I implemented this:
```
            w[15] = w[15].clamp(-1, 1)  # Hard penalty
            
    def stability_after_success(self, state: Tensor, r: Tensor, rating: Tensor) -> Tensor:
        hard_penalty = torch.where(rating == 2, self.w[15], 1)
        pls = self.stability_after_failure(state, r)
        # if hard_penalty <= 0, then SInc = 1 and final S = prev_S * (1 + hard_penalty) + PLS * (1 - (1 + hard_penalty))
        # = prev_S * (1 + hard_penalty) + PLS * -hard_penalty
        # in other words, if hard_penalty <= 0, we take a weighted average of PLS and previous S, otherwise we use new S
        # and hard_penalty acts as a multiplier in SInc

        # if hard_penalty > 0, then the weight of PLS = 0 and SInc > 1, final S = new_S
        # if hard_penalty = 0, then the weight of PLS = 0 and SInc = 1, final S = prev_S
        # if hard_penalty = -1, then the weight of PLS = 1, final S = PLS

        # if hard_penalty = 1, Hard behaves like Good
        # if hard_penalty = -1, Hard behaves like Again
        s_weight = torch.where(hard_penalty <= 0, 1 + hard_penalty, 1)  # the weight of PLS is non-zero only when hard_penalty < 0
        easy_bonus = torch.where(rating == 4, self.w[16], 1)
        new_s = state[:, 0] * (
            1
            + torch.exp(self.w[8])
            * (11 - state[:, 1])
            * torch.pow(state[:, 0], -self.w[9])
            * (torch.exp((1 - r) * self.w[10]) - 1)
            * torch.maximum(hard_penalty, torch.zeros_like(rating))  # Ensure SInc itself is never < 1
            * easy_bonus
        )
        final_s = new_s * s_weight + pls * (1 - s_weight)
        return final_s
```
### w[15] now operates in two distinct regimes: when it's >0, it acts as a multiplier in the SInc formula. When it's <0, it acts as a weight in a weighted average of the previous S and PLS, SInc=1 in this case.

In practice it means that:
1) If w[15]=-1, new S(Hard) = new S(Again)
2) If w[15]=0, new S(Hard) = previous S
3) If w[15]=1, new S(Hard) = new S(Good)

This solves both the problem of the final S being less than PLS or even less than 0, and the problem of S of easy cards dropping harder than S of hard cards. Now this satisfies all intuitively desirable properties while adding more flexibility.
(I tried this before, but messed up the weight so that PLS had the weight of new S and vice versa, hence why it didn't work before)
So, does it work now?
image
...nope, still doesn't work.

**Comment by 1DWalker** [comment]
@Expertium `w[15] = w[15].clamp(-1, 1)` kills the gradient when the value falls outside of (-1, 1) so you might want to use tanh instead.

**Comment by 1DWalker** [comment]
But w[15] can go outside of the range otherwise the clamp doesn't do anything. If w[15] ever goes outside of the range, its value cannot change anymore because the gradient cannot reach it.

**Comment by Expertium** [comment]
Ok, but why would that be a problem here, but not with all the other FSRS parameters, which are also clamped?

**Comment by 1DWalker** [comment]
Where are the other FSRS parameters clamped?

**Comment by Expertium** [comment]
https://github.com/open-spaced-repetition/srs-benchmark/blob/8844f66c006a40ed71a409773468bd4a168b29c2/other.py#L938
image

**Comment by 1DWalker** [comment]
I see, this step falls outside of the computation graph so it works

**Comment by Expertium** [comment]
Also, just to clarify, [here](https://github.com/open-spaced-repetition/fsrs-optimizer/pull/173#issuecomment-2778505539) I didn't want to copy-paste the entire FSRS5ParameterClipper class just to show the difference in *one* line of code.
